platform-sdk/docs/proposals/TSS-Library/tss/TssService.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.tss.api;
+
+import com.swirlds.signaturescheme.api.PairingPrivateKey;
+import com.swirlds.signaturescheme.api.PairingPublicKey;
+import com.swirlds.signaturescheme.api.PairingSignature;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.List;
+
+/**
+ * A Threshold Signature Scheme Service.
+ * <p>
+ * Contract of TSS:
+ * <ul>
+ *     <li>Generate TSSMessages out of PrivateShares</li>
+ *     <li>Verify TSSMessages out of a ParticipantDirectory</li>
+ *     <li>Obtain PrivateShares out of TssMessages for each owned share</li>
+ *     <li>Aggregate PrivateShares</li>
+ *     <li>Obtain PublicShares out of TssMessages for each share</li>
+ *     <li>Aggregate PublicShares</li>
+ *     <li>sign Messages</li>
+ *     <li>verify Signatures</li>
+ *     <li>Aggregate Signatures</li>
+ * </ul>
+ * @implNote an instance of the service would require a source of randomness {@link java.util.Random}, and a{@link com.swirlds.signaturescheme.api.SignatureSchema}
+ */
+public interface TssService {
+
+    /**
+     * Verify that the message is valid.
+     *
+     * @param pendingParticipantDirectory the participant directory used to generate the message
+     * @return true if the message is valid, false otherwise
+     */
+     boolean verifyTssMessage(@NonNull final ParticipantDirectory participantDirectory, @NonNull final TssMessage message);
+
+    /**
+     * Generate a TSS message for a pendingParticipantDirectory, for each private share owned.
+     *
+     * @param pendingParticipantDirectory the pending participant directory that we should generate the message for
+     * @param privateShare       the secret to use for generating new keys
+     * @return a TSSMessage for each owned share.
+     */
+    @NonNull
+    List<TssMessage> generateTssMessages(
```

Comment:
If I understand the proposal correctly, these TssMessages contain encrypted data for each participant in the pending directory. The data is encrypted in each message so that only the intended recipient can decrypt it, therefor each message has a single intended reader. Of course, these messages will all be received by all consensus nodes because of gossip. Without some indication of which message is intended for which recipient, all nodes will have to attempt to decrypt the data in all messages. This seems a bit wasteful. Is this intended? Could it be avoided by including some recipient identifier?

Reply:
> therefore each message has a single intended reader. 

Not quite. Each message has multiple readers (all nodes, even the node that generates the message, are readers for that message).

Each message contains N parts directly mapping to the shares. The ParticipantDirectory tells which part of the message each participant needs to retrieve from one specific message.
```
e.g.:  5 Participants, 10 total shares, 2 shares each.  

Participant 1 owns `share 1` and `share 2`,  it creates 2 TssMessages. 
`Message 1` will split `share 1` into 10 parts, encrypt parts 1 and 2 with its own public key, and the rest with the shareowner's respective public key.

When reading the messages, it first collects a threshold number of valid messages (from all nodes).
Then, using the directory, it knows it needs to read parts 1 and 2 from each collected message.
```

Reply:
ahhh, i see. Thank you for explaining that!